In [1]:
library(wooldridge)
library(stargazer)
library(knitr)


Please cite as: 

 Hlavac, Marek (2018). stargazer: Well-Formatted Regression and Summary Statistics Tables.
 R package version 5.2.2. https://CRAN.R-project.org/package=stargazer 



In [48]:
pacman::p_load(wooldridge,stargazer,broom,tidyverse,ggplot2,
               dplyr,tidyr,moderndive,jtools,sjPlot,dotwhisker,FSA,
               car,QuantPsyc,lmtest,sandwich,tseries, 
              dynlm,prais,clubSandwich,orcutt)

Installing package into ‘/home/jupyter/R/x86_64-pc-linux-gnu-library/3.5’
(as ‘lib’ is unspecified)

orcutt installed


In [5]:
install.packages("prais")

Installing package into 'C:/Users/oldyu/R/win-library/3.5'
(as 'lib' is unspecified)


package 'prais' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\oldyu\AppData\Local\Temp\RtmpMxXRQ0\downloaded_packages


# Example 12.1: The Puerto Rican Minimum Wage



Heteroskedasticity and autocorrelation consistent (HAC) estimation of the covariance matrix of the coefficient estimates in a (generalized) linear regression model.


fm <- lm(y ~ x)

vcovHAC(fm)

vcov(fm)

 clubSandwich ::coef_test

Test all or selected regression coefficients in a fitted model

coef_test reports t-tests for each coefficient estimate in a fitted linear regression model, using a sandwich estimator for the standard errors and a small sample correction for the p-value. The small-sample correction is based on a Satterthwaite approximation or a saddlepoint approximation.

Usage

coef_test(obj, vcov, test = "Satterthwaite", coefs = "All",
  p_values = TRUE, ...)



In [ ]:
data("prminwge")

In [10]:
?prminwge

In [7]:
glimpse(prminwge)

Observations: 38
Variables: 25
$ year     <int> 1950, 1951, 1952, 1953, 1954, 1955, 1956, 1957, 1958, 1959, …
$ avgmin   <dbl> 0.198, 0.209, 0.225, 0.311, 0.313, 0.369, 0.447, 0.488, 0.55…
$ avgwage  <dbl> 0.398, 0.410, 0.421, 0.480, 0.508, 0.547, 0.601, 0.685, 0.71…
$ kaitz    <dbl> 0.155, 0.164, 0.180, 0.229, 0.211, 0.231, 0.257, 0.251, 0.25…
$ avgcov   <dbl> 0.201, 0.207, 0.226, 0.231, 0.224, 0.236, 0.245, 0.244, 0.23…
$ covt     <dbl> 0.29, 0.29, 0.29, 0.29, 0.29, 0.29, 0.29, 0.29, 0.29, 0.29, …
$ mfgwage  <dbl> 0.43, 0.45, 0.48, 0.50, 0.52, 0.57, 0.64, 0.76, 0.83, 0.87, …
$ prdef    <dbl> 0.859, 0.881, 0.953, 0.970, 1.000, 1.003, 1.011, 1.035, 1.08…
$ prepop   <dbl> 0.470, 0.449, 0.434, 0.428, 0.415, 0.419, 0.412, 0.412, 0.39…
$ prepopf  <dbl> 0.470, 0.449, 0.434, 0.428, 0.415, 0.419, 0.412, 0.412, 0.39…
$ prgnp    <dbl> 878.7, 925.0, 1015.9, 1081.3, 1104.4, 1138.5, 1185.1, 1221.8…
$ prunemp  <dbl> 15.4, 16.0, 14.8, 14.5, 15.3, 13.2, 13.3, 12.8, 14.2, 13.3, …
$ usgnp    <dbl> 1203

In [11]:


lm.12.1 <- lm(lprepop  ~ lmincov + lusgnp + lprgnp+t, data = prminwge)
summary(lm.12.1 )


Call:
lm(formula = lprepop ~ lmincov + lusgnp + lprgnp + t, data = prminwge)

Residuals:
      Min        1Q    Median        3Q       Max 
-0.054679 -0.023653 -0.004039  0.018638  0.076947 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) -6.663432   1.257831  -5.298 7.67e-06 ***
lmincov     -0.212261   0.040152  -5.286 7.92e-06 ***
lusgnp       0.486046   0.221983   2.190   0.0357 *  
lprgnp       0.285239   0.080492   3.544   0.0012 ** 
t           -0.026663   0.004627  -5.763 1.94e-06 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.03277 on 33 degrees of freedom
Multiple R-squared:  0.8892,	Adjusted R-squared:  0.8758 
F-statistic: 66.23 on 4 and 33 DF,  p-value: 2.677e-15


In [35]:
vcovHAC(lm.12.1)

(Intercept)  lmincov       lusgnp       lprgnp        t            
(Intercept)  2.841552038  3.446714e-02 -0.510891872  0.1280964927  8.231351e-03
lmincov      0.034467136  2.122299e-03 -0.005290986  0.0011158990  2.580249e-05
lusgnp      -0.510891872 -5.290986e-03  0.096655218 -0.0278272361 -1.422111e-03
lprgnp       0.128096493  1.115899e-03 -0.027827236  0.0107102031  2.870062e-04
t            0.008231351  2.580249e-05 -0.001422111  0.0002870062  2.948585e-05

In [12]:
coeftest( lm.12.1 , vcov = vcovHAC(lm.12.1))



t test of coefficients:

              Estimate Std. Error t value  Pr(>|t|)    
(Intercept) -6.6634319  1.6856904 -3.9529 0.0003845 ***
lmincov     -0.2122612  0.0460684 -4.6075 5.835e-05 ***
lusgnp       0.4860463  0.3108942  1.5634 0.1275031    
lprgnp       0.2852386  0.1034901  2.7562 0.0094496 ** 
t           -0.0266633  0.0054301 -4.9103 2.402e-05 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1


In [54]:
library(dynlm);library(lmtest);library(sandwich)
data(prminwge, package='wooldridge')

tsdata <- ts(prminwge, start=1950)

# OLS regression
reg<-dynlm(log(prepop)~log(mincov)+log(prgnp)+log(usgnp)+trend(tsdata), 
                                                          data=tsdata )
# results with usual SE
coeftest(reg)
# results with HAC SE
coeftest(reg, vcovHAC)


t test of coefficients:

                Estimate Std. Error t value  Pr(>|t|)    
(Intercept)   -6.6634416  1.2578286 -5.2976 7.667e-06 ***
log(mincov)   -0.2122612  0.0401523 -5.2864 7.924e-06 ***
log(prgnp)     0.2852380  0.0804921  3.5437  0.001203 ** 
log(usgnp)     0.4860483  0.2219825  2.1896  0.035731 *  
trend(tsdata) -0.0266633  0.0046267 -5.7629 1.940e-06 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



t test of coefficients:

                Estimate Std. Error t value  Pr(>|t|)    
(Intercept)   -6.6634416  1.6856885 -3.9529 0.0003845 ***
log(mincov)   -0.2122612  0.0460684 -4.6075 5.835e-05 ***
log(prgnp)     0.2852380  0.1034901  2.7562 0.0094497 ** 
log(usgnp)     0.4860483  0.3108940  1.5634 0.1275013    
trend(tsdata) -0.0266633  0.0054301 -4.9103 2.402e-05 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1


In [64]:
data("prminwge")

lm.12.2 <- lm(lprepop ~ lmincov + lusgnp + lprgnp + t, data = prminwge)
lm.12.2.res <- lm.12.2$res
lm.12.2.res_1 <- c(NA, lm.12.2$res[-length(lm.12.2.res)])
lm.12.2.test <- lm(lm.12.2.res ~ lmincov + lusgnp + lprgnp + t + lm.12.2.res_1, data = prminwge)
summary(lm.12.2.test)


Call:
lm(formula = lm.12.2.res ~ lmincov + lusgnp + lprgnp + t + lm.12.2.res_1, 
    data = prminwge)

Residuals:
      Min        1Q    Median        3Q       Max 
-0.041317 -0.018004 -0.004599  0.012378  0.067226 

Coefficients:
               Estimate Std. Error t value Pr(>|t|)   
(Intercept)   -0.850772   1.092691  -0.779  0.44212   
lmincov        0.037500   0.035212   1.065  0.29511   
lusgnp         0.203932   0.195159   1.045  0.30412   
lprgnp        -0.078466   0.070524  -1.113  0.27443   
t             -0.003466   0.004074  -0.851  0.40134   
lm.12.2.res_1  0.480509   0.166444   2.887  0.00703 **
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.02755 on 31 degrees of freedom
  (1 observation deleted due to missingness)
Multiple R-squared:  0.2424,	Adjusted R-squared:  0.1202 
F-statistic: 1.983 on 5 and 31 DF,  p-value: 0.1089


In [65]:
#For comparison the test with strict exogenous regressors

summary(lm(lm.12.2.res ~ lm.12.2.res_1))


Call:
lm(formula = lm.12.2.res ~ lm.12.2.res_1)

Residuals:
      Min        1Q    Median        3Q       Max 
-0.043815 -0.024235 -0.002278  0.015155  0.063480 

Coefficients:
                Estimate Std. Error t value Pr(>|t|)  
(Intercept)   -0.0008953  0.0044883  -0.199   0.8431  
lm.12.2.res_1  0.4173219  0.1589351   2.626   0.0127 *
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.02723 on 35 degrees of freedom
  (1 observation deleted due to missingness)
Multiple R-squared:  0.1646,	Adjusted R-squared:  0.1407 
F-statistic: 6.895 on 1 and 35 DF,  p-value: 0.01274


# Example 12.2: Static Phillips curve

In [3]:
data("phillips")
lm.12.1.1 <- lm(inf ~ unem, data = phillips)
res.static <- lm.12.1.1$res
res.static_1 <- c(NA, lm.12.1.1$res[-(length(res.static))])
lm.12.1.1.test <- lm(res.static ~ res.static_1)
summary(lm.12.1.1.test)


Call:
lm(formula = res.static ~ res.static_1)

Residuals:
   Min     1Q Median     3Q    Max 
-8.047 -1.104 -0.248  1.028  6.684 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept)   -0.1118     0.3180  -0.352    0.727    
res.static_1   0.5725     0.1084   5.283 2.43e-06 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.358 on 53 degrees of freedom
  (1 observation deleted due to missingness)
Multiple R-squared:  0.345,	Adjusted R-squared:  0.3326 
F-statistic: 27.91 on 1 and 53 DF,  p-value: 2.43e-06


In [6]:
library(dynlm);library(lmtest)
data(phillips, package='wooldridge')

# Define Yearly time series beginning in 1948
tsdata <- ts(phillips, start=1948)

# Estimation of static Phillips curve:
reg.s <- dynlm( inf ~ unem, data=tsdata, end=1996)
# residuals and AR(1) test:
residual.s <- resid(reg.s)
coeftest( dynlm(residual.s ~ L(residual.s)) )

# Same with expectations-augmented Phillips curve:
reg.ea <- dynlm( d(inf) ~ unem, data=tsdata, end=1996)
residual.ea <- resid(reg.ea)
coeftest( dynlm(residual.ea ~ L(residual.ea)) )


t test of coefficients:

              Estimate Std. Error t value  Pr(>|t|)    
(Intercept)   -0.11340    0.35940 -0.3155    0.7538    
L(residual.s)  0.57297    0.11613  4.9337 1.098e-05 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



t test of coefficients:

                Estimate Std. Error t value Pr(>|t|)
(Intercept)     0.194166   0.300384  0.6464   0.5213
L(residual.ea) -0.035593   0.123891 -0.2873   0.7752


# 12.3b The Durbin-Watson Test under Classical Assumption

In [13]:
library(dynlm);library(lmtest)
data(phillips, package='wooldridge')

tsdata <- ts(phillips, start=1948)

# Estimation of both Phillips curve models:
reg.s <- dynlm( inf ~ unem, data=tsdata, end=1996)
reg.ea <- dynlm( d(inf) ~ unem, data=tsdata, end=1996)

# DW tests
dwtest(reg.s)
dwtest(reg.ea)


	Durbin-Watson test

data:  reg.s
DW = 0.8027, p-value = 7.552e-07
alternative hypothesis: true autocorrelation is greater than 0



	Durbin-Watson test

data:  reg.ea
DW = 1.7696, p-value = 0.1783
alternative hypothesis: true autocorrelation is greater than 0


In [62]:
# from package lmtest
dwtest(inf ~ unem, data = phillips)


	Durbin-Watson test

data:  inf ~ unem
DW = 0.80148, p-value = 1.486e-07
alternative hypothesis: true autocorrelation is greater than 0


In [63]:
dwtest(I(inf - inf_1) ~ unem, data = phillips)


	Durbin-Watson test

data:  I(inf - inf_1) ~ unem
DW = 1.771, p-value = 0.1673
alternative hypothesis: true autocorrelation is greater than 0


# Example 12.3: Testing for AR(1) SC in the minimum wage equation

In [15]:
uthat = resid(lm.12.1 )

In [41]:
# prminwge

In [42]:
# lm.12.3 = 
# prminwge[-1,]

In [28]:
tsprminwge <- ts(prminwge, start=1950)

In [30]:
res.static <- lm.12.1$res
res.static_1 <- c(NA, lm.12.1$res[-(length(res.static))])

In [33]:
coeftest( lm(res.static  ~ res.static_1 + lmincov + lusgnp + lprgnp+t, data=prminwge))


t test of coefficients:

               Estimate Std. Error t value Pr(>|t|)   
(Intercept)  -0.8507721  1.0926905 -0.7786 0.442117   
res.static_1  0.4805093  0.1664442  2.8869 0.007029 **
lmincov       0.0375001  0.0352123  1.0650 0.295114   
lusgnp        0.2039325  0.1951588  1.0450 0.304123   
lprgnp       -0.0784656  0.0705240 -1.1126 0.274430   
t            -0.0034662  0.0040736 -0.8509 0.401345   
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1


In [78]:
data("prminwge")

lm.12.7.1 <- lm(lprepop ~ lmincov + lusgnp + lprgnp + t, data = prminwge)
summary(lm.12.7.1)


Call:
lm(formula = lprepop ~ lmincov + lusgnp + lprgnp + t, data = prminwge)

Residuals:
      Min        1Q    Median        3Q       Max 
-0.054679 -0.023653 -0.004039  0.018638  0.076947 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) -6.663432   1.257831  -5.298 7.67e-06 ***
lmincov     -0.212261   0.040152  -5.286 7.92e-06 ***
lusgnp       0.486046   0.221983   2.190   0.0357 *  
lprgnp       0.285239   0.080492   3.544   0.0012 ** 
t           -0.026663   0.004627  -5.763 1.94e-06 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.03277 on 33 degrees of freedom
Multiple R-squared:  0.8892,	Adjusted R-squared:  0.8758 
F-statistic: 66.23 on 4 and 33 DF,  p-value: 2.677e-15


In [79]:
# Still looking for the SC/heteroskedasticiy robust se.

pw_12_7_1 <- prais_winsten(lprepop ~ lmincov + lusgnp + lprgnp + t, data = prminwge)

Iteration 0: rho = 0
Iteration 1: rho = 0.4197
Iteration 2: rho = 0.5325
Iteration 3: rho = 0.5796
Iteration 4: rho = 0.5999
Iteration 5: rho = 0.6086
Iteration 6: rho = 0.6123
Iteration 7: rho = 0.6139
Iteration 8: rho = 0.6146
Iteration 9: rho = 0.6149
Iteration 10: rho = 0.615
Iteration 11: rho = 0.6151
Iteration 12: rho = 0.6151
Iteration 13: rho = 0.6151
Iteration 14: rho = 0.6151
Iteration 15: rho = 0.6151
Iteration 16: rho = 0.6151


In [80]:
summary(pw_12_7_1)


Call:
prais_winsten(formula = lprepop ~ lmincov + lusgnp + lprgnp + 
    t, data = prminwge)

Residuals:
      Min        1Q    Median        3Q       Max 
-0.079574 -0.028121 -0.004816  0.005918  0.075978 

AR(1) coefficient rho after 16 Iterations: 0.6151

Coefficients:
             Estimate Std. Error t value Pr(>|t|)   
(Intercept) -4.652853   1.376470  -3.380  0.00188 **
lmincov     -0.147711   0.045842  -3.222  0.00286 **
lusgnp       0.255711   0.231750   1.103  0.27784   
lprgnp       0.251383   0.116462   2.158  0.03826 * 
t           -0.020502   0.005856  -3.501  0.00135 **
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.02852 on 33 degrees of freedom
Multiple R-squared:  0.7509,	Adjusted R-squared:  0.7207 
F-statistic: 24.87 on 4 and 33 DF,  p-value: 1.47e-09

Durbin-Watson statistic (original): 1.014 
Durbin-Watson statistic (transformed): 1.736


In [ ]:
library(dynlm);library(car);library(lmtest)
data(barium, package='wooldridge')

tsdata <- ts(barium, start=c(1978,2), frequency=12)

reg <- dynlm(log(chnimp)~log(chempi)+log(gas)+log(rtwex)+
                                  befile6+affile6+afdec6, data=tsdata )

# Pedestrian test: 
residual <- resid(reg)
resreg <- dynlm(residual ~ L(residual)+L(residual,2)+L(residual,3)+
                           log(chempi)+log(gas)+log(rtwex)+befile6+
                                          affile6+afdec6, data=tsdata )
linearHypothesis(resreg, 
                 c("L(residual)","L(residual, 2)","L(residual, 3)"))

# Automatic test:
bgtest(reg, order=3, type="F")

# Example 12.4: Testing for AR(3) SC


Automatic test: bgtest


Basically you need the sandwich package, which computes robust covariance matrix estimators. You also need some way to use the variance estimator in a linear model, and the lmtest package is the solution. You will not get the same results as Stata, however, unless you use the HC1 estimator; the default is HC3, for reasons explained in ?vcovHC.

In [38]:
library(dynlm);library(car);library(lmtest)
data(barium, package='wooldridge')

tsdata <- ts(barium, start=c(1978,2), frequency=12)

lm.10.22=reg <- dynlm(log(chnimp)~log(chempi)+log(gas)+log(rtwex)+
                                  befile6+affile6+afdec6, data=tsdata )

# Pedestrian test: 
residual <- resid(reg)
resreg <- dynlm(residual ~ L(residual)+L(residual,2)+L(residual,3)+
                           log(chempi)+log(gas)+log(rtwex)+befile6+
                                          affile6+afdec6, data=tsdata )
linearHypothesis(resreg, 
                 c("L(residual)","L(residual, 2)","L(residual, 3)"))

# Automatic test:
bgtest(reg, order=3, type="F")

Res.Df RSS      Df Sum of Sq F        Pr(>F)     
1 121    43.39419 NA       NA        NA          NA
2 118    38.39367  3 5.000522  5.122907 0.002289803


	Breusch-Godfrey test for serial correlation of order up to 3

data:  reg
LM test = 5.1247, df1 = 3, df2 = 121, p-value = 0.002264


In [66]:
data("barium")

lm.12.3 <- lm(lchnimp ~ lchempi + lgas + lrtwex + befile6 + affile6 + afdec6, data = barium)

lm.12.3.res <- lm.12.3$res
lm.12.3.res_1 <- c(NA,lm.12.3$res[1:(length(lm.12.3.res)-1)])
lm.12.3.res_2 <- c(NA,NA,lm.12.3$res[1:(length(lm.12.3.res)-2)])
lm.12.3.res_3 <- c(NA,NA,NA,lm.12.3$res[1:(length(lm.12.3.res)-3)])

data <- cbind(barium, lm.12.3.res)
data <- cbind(data, lm.12.3.res_1)
data <- cbind(data, lm.12.3.res_2)
data <- cbind(data, lm.12.3.res_3)

lm.12.3.test <- lm(lm.12.3.res ~ lchempi + lgas + lrtwex + befile6 + affile6 + afdec6 +
 lm.12.3.res_1 + lm.12.3.res_2 + lm.12.3.res_3, data = data)

lm.12.3.test.res <- lm(lm.12.3.res ~ lchempi + lgas + lrtwex + befile6 + affile6 + afdec6,
 data = data[-(1:3),]) # Drop first 3 observations to make test statistics comparable

anova(lm.12.3.test, lm.12.3.test.res)

Res.Df RSS      Df Sum of Sq F        Pr(>F)     
1 118    38.39366 NA        NA       NA          NA
2 121    43.39419 -3 -5.000528 5.122914 0.002289782

In [67]:
#Seasonal autocorrelation

data <- cbind(data, lm.12.3.res_12 = c(rep(NA, 12), lm.12.3$res[1:(length(lm.12.3.res) - 12)]))
summary(lm(lm.12.3.res ~ lm.12.3.res_12, data = data))


Call:
lm(formula = lm.12.3.res ~ lm.12.3.res_12, data = data)

Residuals:
     Min       1Q   Median       3Q      Max 
-1.96217 -0.33069  0.03732  0.36389  1.46367 

Coefficients:
               Estimate Std. Error t value Pr(>|t|)  
(Intercept)     0.03789    0.05047   0.751    0.454  
lm.12.3.res_12 -0.18744    0.08425  -2.225    0.028 *
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.5506 on 117 degrees of freedom
  (12 observations deleted due to missingness)
Multiple R-squared:  0.04059,	Adjusted R-squared:  0.03239 
F-statistic:  4.95 on 1 and 117 DF,  p-value: 0.02801


In [68]:
#Include regressors

summary(lm(lm.12.3.res ~ lchempi + lgas + lrtwex + befile6 + affile6 + afdec6 +
 lm.12.3.res_12, data = data))


Call:
lm(formula = lm.12.3.res ~ lchempi + lgas + lrtwex + befile6 + 
    affile6 + afdec6 + lm.12.3.res_12, data = data)

Residuals:
    Min      1Q  Median      3Q     Max 
-2.0259 -0.3467  0.0158  0.3770  1.3676 

Coefficients:
                Estimate Std. Error t value Pr(>|t|)  
(Intercept)    -28.57195   22.64780  -1.262   0.2097  
lchempi         -0.92803    0.54647  -1.698   0.0923 .
lgas             1.50251    1.01497   1.480   0.1416  
lrtwex          -0.27439    0.38127  -0.720   0.4732  
befile6          0.04437    0.24581   0.180   0.8571  
affile6          0.06253    0.24600   0.254   0.7998  
afdec6           0.17231    0.27038   0.637   0.5253  
lm.12.3.res_12  -0.17002    0.08670  -1.961   0.0524 .
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.5541 on 111 degrees of freedom
  (12 observations deleted due to missingness)
Multiple R-squared:  0.07811,	Adjusted R-squared:  0.01997 
F-statistic: 1.344 on 7 and 111 DF,  p-v

# Example 12.4.1: Testing for AR(3) SC

`

In [36]:
# Automatic test:
bgtest(reg, order=12, type="F")


	Breusch-Godfrey test for serial correlation of order up to 12

data:  reg
LM test = 2.0865, df1 = 12, df2 = 112, p-value = 0.02319


In [39]:

resreg <- dynlm(residual ~ L(residual,12), data=tsdata )

summary(resreg)



Time series regression with "ts" data:
Start = 1979(2), End = 1988(12)

Call:
dynlm(formula = residual ~ L(residual, 12), data = tsdata)

Residuals:
     Min       1Q   Median       3Q      Max 
-1.96217 -0.33069  0.03732  0.36389  1.46367 

Coefficients:
                Estimate Std. Error t value Pr(>|t|)  
(Intercept)      0.03789    0.05047   0.751    0.454  
L(residual, 12) -0.18744    0.08425  -2.225    0.028 *
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.5506 on 117 degrees of freedom
Multiple R-squared:  0.04059,	Adjusted R-squared:  0.03239 
F-statistic:  4.95 on 1 and 117 DF,  p-value: 0.02801


In [44]:
# Pedestrian test: 
residual <- resid(reg)
resreg <- dynlm(residual ~ L(residual,12)+
                           log(chempi)+log(gas)+log(rtwex)+befile6+
                                          affile6+afdec6, data=tsdata )

summary(resreg)
# linearHypothesis(resreg, 
#                  c("L(residual)","L(residual, 12)","L(residual, 3)"))


Time series regression with "ts" data:
Start = 1979(2), End = 1988(12)

Call:
dynlm(formula = residual ~ L(residual, 12) + log(chempi) + log(gas) + 
    log(rtwex) + befile6 + affile6 + afdec6, data = tsdata)

Residuals:
    Min      1Q  Median      3Q     Max 
-2.0259 -0.3467  0.0158  0.3770  1.3676 

Coefficients:
                 Estimate Std. Error t value Pr(>|t|)  
(Intercept)     -28.57203   22.64780  -1.262   0.2097  
L(residual, 12)  -0.17002    0.08670  -1.961   0.0524 .
log(chempi)      -0.92803    0.54647  -1.698   0.0923 .
log(gas)          1.50252    1.01497   1.480   0.1416  
log(rtwex)       -0.27439    0.38127  -0.720   0.4732  
befile6           0.04437    0.24581   0.180   0.8571  
affile6           0.06253    0.24600   0.254   0.7998  
afdec6            0.17231    0.27038   0.637   0.5253  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.5541 on 111 degrees of freedom
Multiple R-squared:  0.07811,	Adjusted R-squared:  

In [50]:
coeftest(resreg, vcov =  'HC1') #vcovHAC(resreg) )

Warning message in diag(se):
“NAs introduced by coercion”

ERROR: Error in diag(se): invalid 'nrow' value (too large or NA)


In [49]:
library(dynlm);library(car);library(orcutt)
data(barium, package='wooldridge')

tsdata <- ts(barium, start=c(1978,2), frequency=12)

# OLS estimation
olsres <- dynlm(log(chnimp)~log(chempi)+log(gas)+log(rtwex)+
      befile6+affile6+afdec6, data=tsdata)

# Cochrane-Orcutt estimation
cochrane.orcutt(olsres)

Cochrane-orcutt estimation for first order autocorrelation 
 
Call:
dynlm(formula = log(chnimp) ~ log(chempi) + log(gas) + log(rtwex) + 
    befile6 + affile6 + afdec6, data = tsdata)

 number of interaction: 8
 rho 0.293362

Durbin-Watson statistic 
(original):    1.45841 , p-value: 1.688e-04
(transformed): 2.06330 , p-value: 4.91e-01
 
 coefficients: 
(Intercept) log(chempi)    log(gas)  log(rtwex)     befile6     affile6 
 -37.322241    2.947434    1.054858    1.136918   -0.016372   -0.033082 
     afdec6 
  -0.577158 

# Example 12.5: Prais-Winsten Estimation in the Event Study

In [3]:
data("barium")

barium_model <- lm(lchnimp ~ lchempi + lgas + lrtwex + befile6 + affile6 + afdec6, 
                   data = barium)


In [7]:
stargazer(type = "text", barium_model, single.row = TRUE, header = FALSE, digits=5)



                        Dependent variable:    
                    ---------------------------
                              lchnimp          
-----------------------------------------------
lchempi                3.11719*** (0.47920)    
lgas                     0.19635 (0.90662)     
lrtwex                  0.98302** (0.40015)    
befile6                  0.05957 (0.26097)     
affile6                 -0.03241 (0.26430)     
afdec6                  -0.56524* (0.28584)    
Constant               -17.80300 (21.04537)    
-----------------------------------------------
Observations                    131            
R2                            0.30486          
Adjusted R2                   0.27123          
Residual Std. Error     0.59735 (df = 124)     
F Statistic          9.06365*** (df = 6; 124)  
Note:               *p<0.1; **p<0.05; ***p<0.01


In [52]:
# Load the prais package, use the prais_winsten function to estimate.

library(prais)
barium_prais_winsten <- prais_winsten(lchnimp ~ lchempi + lgas + lrtwex + befile6 + affile6 + afdec6, data = barium)


Iteration 0: rho = 0
Iteration 1: rho = 0.2708
Iteration 2: rho = 0.291
Iteration 3: rho = 0.293
Iteration 4: rho = 0.2932
Iteration 5: rho = 0.2932
Iteration 6: rho = 0.2932
Iteration 7: rho = 0.2932


In [53]:
summary(barium_prais_winsten)


Call:
prais_winsten(formula = lchnimp ~ lchempi + lgas + lrtwex + befile6 + 
    affile6 + afdec6, data = barium)

Residuals:
     Min       1Q   Median       3Q      Max 
-1.99386 -0.32219  0.03747  0.40226  1.50281 

AR(1) coefficient rho after 7 Iterations: 0.2932

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) -37.07770   22.77830  -1.628   0.1061    
lchempi       2.94095    0.63284   4.647 8.46e-06 ***
lgas          1.04638    0.97734   1.071   0.2864    
lrtwex        1.13279    0.50666   2.236   0.0272 *  
befile6      -0.01648    0.31938  -0.052   0.9589    
affile6      -0.03316    0.32181  -0.103   0.9181    
afdec6       -0.57681    0.34199  -1.687   0.0942 .  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.5733 on 124 degrees of freedom
Multiple R-squared:  0.2021,	Adjusted R-squared:  0.1635 
F-statistic: 5.235 on 6 and 124 DF,  p-value: 7.764e-05

Durbin-Watson statistic (original): 1.458 


In [51]:
library(dynlm);library(car);library(orcutt)
data(barium, package='wooldridge')

tsdata <- ts(barium, start=c(1978,2), frequency=12)

# OLS estimation
olsres <- dynlm(log(chnimp)~log(chempi)+log(gas)+log(rtwex)+
      befile6+affile6+afdec6, data=tsdata)

# Cochrane-Orcutt estimation
cochrane.orcutt(olsres)

Cochrane-orcutt estimation for first order autocorrelation 
 
Call:
dynlm(formula = log(chnimp) ~ log(chempi) + log(gas) + log(rtwex) + 
    befile6 + affile6 + afdec6, data = tsdata)

 number of interaction: 8
 rho 0.293362

Durbin-Watson statistic 
(original):    1.45841 , p-value: 1.688e-04
(transformed): 2.06330 , p-value: 4.91e-01
 
 coefficients: 
(Intercept) log(chempi)    log(gas)  log(rtwex)     befile6     affile6 
 -37.322241    2.947434    1.054858    1.136918   -0.016372   -0.033082 
     afdec6 
  -0.577158 

# Example 12.5 Static Philips Curves

In [58]:
library(dynlm);library(lmtest)
data(phillips, package='wooldridge')

# Define Yearly time series beginning in 1948
tsdata <- ts(phillips, start=1948)

# Estimation of static Phillips curve:
reg.s <- dynlm( inf ~ unem, data=tsdata, end=2006)
summary(reg.s)
# residuals and AR(1) test:
residual.s <- resid(reg.s)
coeftest( dynlm(residual.s ~ L(residual.s)) )

# Same with expectations-augmented Phillips curve:
reg.ea <- dynlm( d(inf) ~ unem, data=tsdata, end=2006)
summary(reg.ea)
residual.ea <- resid(reg.ea)
coeftest( dynlm(residual.ea ~ L(residual.ea)) )


Time series regression with "ts" data:
Start = 1948, End = 2003

Call:
dynlm(formula = inf ~ unem, data = tsdata, end = 2006)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.2176 -1.7812 -0.6659  1.1473  8.8795 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)  
(Intercept)   1.0536     1.5480   0.681   0.4990  
unem          0.5024     0.2656   1.892   0.0639 .
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.972 on 54 degrees of freedom
Multiple R-squared:  0.06215,	Adjusted R-squared:  0.04479 
F-statistic: 3.579 on 1 and 54 DF,  p-value: 0.06389



t test of coefficients:

              Estimate Std. Error t value Pr(>|t|)    
(Intercept)   -0.11181    0.31799 -0.3516   0.7265    
L(residual.s)  0.57247    0.10835  5.2833 2.43e-06 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



Time series regression with "ts" data:
Start = 1949, End = 2003

Call:
dynlm(formula = d(inf) ~ unem, data = tsdata, end = 2006)

Residuals:
    Min      1Q  Median      3Q     Max 
-9.0741 -0.9241  0.0189  0.8606  5.4800 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)  
(Intercept)   2.8282     1.2249   2.309   0.0249 *
unem         -0.5176     0.2090  -2.476   0.0165 *
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.307 on 53 degrees of freedom
Multiple R-squared:  0.1037,	Adjusted R-squared:  0.08679 
F-statistic: 6.132 on 1 and 53 DF,  p-value: 0.0165



t test of coefficients:

                Estimate Std. Error t value Pr(>|t|)
(Intercept)     0.167446   0.265478  0.6307   0.5310
L(residual.ea) -0.032697   0.116358 -0.2810   0.7798


In [61]:
summary(dynlm( d(inf) ~ d(unem), data=tsdata, end=2006))


Time series regression with "ts" data:
Start = 1949, End = 2003

Call:
dynlm(formula = d(inf) ~ d(unem), data = tsdata, end = 2006)

Residuals:
    Min      1Q  Median      3Q     Max 
-7.4790 -0.9441  0.1384  1.0889  5.4551 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)   
(Intercept) -0.07214    0.30584  -0.236  0.81443   
d(unem)     -0.83281    0.28984  -2.873  0.00583 **
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.267 on 53 degrees of freedom
Multiple R-squared:  0.1348,	Adjusted R-squared:  0.1185 
F-statistic: 8.256 on 1 and 53 DF,  p-value: 0.005831


In [56]:
# Cochrane-Orcutt estimation
cochrane.orcutt(reg.s )

Cochrane-orcutt estimation for first order autocorrelation 
 
Call:
dynlm(formula = inf ~ unem, data = tsdata, end = 1996)

 number of interaction: 13
 rho 0.774051

Durbin-Watson statistic 
(original):    0.80270 , p-value: 7.552e-07
(transformed): 1.59363 , p-value: 7.447e-02
 
 coefficients: 
(Intercept)        unem 
   7.583459   -0.665336 

In [57]:
# Cochrane-Orcutt estimation
cochrane.orcutt(reg.ea)

Cochrane-orcutt estimation for first order autocorrelation 
 
Call:
dynlm(formula = d(inf) ~ unem, data = tsdata, end = 1996)

 number of interaction: 5
 rho -0.037924

Durbin-Watson statistic 
(original):    1.76965 , p-value: 1.783e-01
(transformed): 1.84956 , p-value: 2.612e-01
 
 coefficients: 
(Intercept)        unem 
   3.102203   -0.522611 

In [71]:
#First column

lm.12.5 <- lm(inf ~ unem, data = phillips)
summary(lm.12.5)
# Second column
pw_2 <- prais_winsten(inf ~ unem, data=phillips)


Call:
lm(formula = inf ~ unem, data = phillips)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.2176 -1.7812 -0.6659  1.1473  8.8795 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)  
(Intercept)   1.0536     1.5480   0.681   0.4990  
unem          0.5024     0.2656   1.892   0.0639 .
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.972 on 54 degrees of freedom
Multiple R-squared:  0.06215,	Adjusted R-squared:  0.04479 
F-statistic: 3.579 on 1 and 54 DF,  p-value: 0.06389


Iteration 0: rho = 0
Iteration 1: rho = 0.5721
Iteration 2: rho = 0.735
Iteration 3: rho = 0.7792
Iteration 4: rho = 0.7871
Iteration 5: rho = 0.7883
Iteration 6: rho = 0.7885
Iteration 7: rho = 0.7885
Iteration 8: rho = 0.7885
Iteration 9: rho = 0.7885


In [70]:
summary(pw_2)


Call:
prais_winsten(formula = inf ~ unem, data = phillips)

Residuals:
   Min     1Q Median     3Q    Max 
-5.258 -2.447 -1.073  1.463 10.570 

AR(1) coefficient rho after 9 Iterations: 0.7885

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   7.9994     2.0483   3.905 0.000264 ***
unem         -0.7140     0.2898  -2.464 0.016965 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.138 on 54 degrees of freedom
Multiple R-squared:  0.1345,	Adjusted R-squared:  0.1185 
F-statistic: 8.393 on 1 and 54 DF,  p-value: 0.00543

Durbin-Watson statistic (original): 0.8015 
Durbin-Watson statistic (transformed): 1.914


In [59]:
phillips_prais_winsten <- prais_winsten(inf ~ unem, data=tsdata)


Iteration 0: rho = 0
Iteration 1: rho = 0.5721
Iteration 2: rho = 0.735
Iteration 3: rho = 0.7792
Iteration 4: rho = 0.7871
Iteration 5: rho = 0.7883
Iteration 6: rho = 0.7885
Iteration 7: rho = 0.7885
Iteration 8: rho = 0.7885
Iteration 9: rho = 0.7885


In [60]:
summary(phillips_prais_winsten)


Call:
prais_winsten(formula = inf ~ unem, data = tsdata)

Residuals:
   Min     1Q Median     3Q    Max 
-5.258 -2.447 -1.073  1.463 10.570 

AR(1) coefficient rho after 9 Iterations: 0.7885

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   7.9994     2.0483   3.905 0.000264 ***
unem         -0.7140     0.2898  -2.464 0.016965 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.138 on 54 degrees of freedom
Multiple R-squared:  0.1345,	Adjusted R-squared:  0.1185 
F-statistic: 8.393 on 1 and 54 DF,  p-value: 0.00543

Durbin-Watson statistic (original): 0.8015 
Durbin-Watson statistic (transformed): 1.914


# Example 12.7 Differencing the interest rate equation

In [72]:
data("intdef")

summary(lm.12.7.1 <- lm(i3 ~ inf + def, data = intdef))


Call:
lm(formula = i3 ~ inf + def, data = intdef)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.9948 -1.1694  0.1959  0.9602  4.7224 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  1.73327    0.43197   4.012  0.00019 ***
inf          0.60587    0.08213   7.376 1.12e-09 ***
def          0.51306    0.11838   4.334 6.57e-05 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1.843 on 53 degrees of freedom
Multiple R-squared:  0.6021,	Adjusted R-squared:  0.5871 
F-statistic: 40.09 on 2 and 53 DF,  p-value: 2.483e-11


In [73]:
res <- lm.12.7.1$residuals
res_1 <- c(NA, res[-length(res)])
summary(lm(res ~ -1 + res_1))


Call:
lm(formula = res ~ -1 + res_1)

Residuals:
    Min      1Q  Median      3Q     Max 
-2.8010 -0.6616  0.0179  0.8721  4.5901 

Coefficients:
      Estimate Std. Error t value Pr(>|t|)    
res_1   0.6229     0.1087    5.73 4.61e-07 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1.397 on 54 degrees of freedom
  (1 observation deleted due to missingness)
Multiple R-squared:  0.3781,	Adjusted R-squared:  0.3666 
F-statistic: 32.83 on 1 and 54 DF,  p-value: 4.612e-07


In [74]:
lm.12.7.2 <- lm(I(diff(i3)) ~ I(diff(inf)) + I(diff(def)), data = intdef)
summary(lm.12.7.2)


Call:
lm(formula = I(diff(i3)) ~ I(diff(inf)) + I(diff(def)), data = intdef)

Residuals:
     Min       1Q   Median       3Q      Max 
-2.53316 -0.96248 -0.08658  0.75884  2.93846 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)
(Intercept)   0.04177    0.17139   0.244    0.808
I(diff(inf))  0.14949    0.09216   1.622    0.111
I(diff(def)) -0.18132    0.14768  -1.228    0.225

Residual standard error: 1.265 on 52 degrees of freedom
Multiple R-squared:  0.1763,	Adjusted R-squared:  0.1446 
F-statistic: 5.566 on 2 and 52 DF,  p-value: 0.006451


In [75]:
#Correlation between i3t and i3t−1

cor(intdef$i3, c(NA, intdef$i3[-length(intdef$i3)]), use = "pairwise.complete.obs")

[1] 0.8845033

In [77]:
#Regression of et on et−1

res <- lm.12.7.2$residuals
res_1 <- res_1 <- c(NA, res[-length(res)])
summary(lm(res ~ -1 + res_1))


Call:
lm(formula = res ~ -1 + res_1)

Residuals:
    Min      1Q  Median      3Q     Max 
-2.7439 -0.9833 -0.1421  0.7773  2.8453 

Coefficients:
      Estimate Std. Error t value Pr(>|t|)
res_1  0.07172    0.13305   0.539    0.592

Residual standard error: 1.212 on 53 degrees of freedom
  (1 observation deleted due to missingness)
Multiple R-squared:  0.005454,	Adjusted R-squared:  -0.01331 
F-statistic: 0.2906 on 1 and 53 DF,  p-value: 0.5921


# Example 12.8: Heteroskedasticity and the Efficient Markets Hypothesis
These are Wednesday closing prices of value-weighted NYSE average, available in many publications. Wooldridge does not recall the particular source used when he collected these data at MIT, but notes probably the easiest way to get similar data is to go to the NYSE web site, www.nyse.com.

$$return_t = \beta_0 + \beta_1return_{t-1} + \mu_t$$

In [10]:
data("nyse")
?nyse 
return_AR1 <-lm(return ~ return_1, data = nyse)

$$\hat{\mu^2_t} = \beta_0 + \beta_1return_{t-1} + residual_t$$

In [11]:
return_mu <- residuals(return_AR1)
mu2_hat_model <- lm(return_mu^2 ~ return_1, data = return_AR1$model)
stargazer(type = "text",return_AR1, mu2_hat_model,  single.row = TRUE, header = FALSE, digits=5)


                                          Dependent variable:           
                               -----------------------------------------
                                     return             return_mu2      
                                       (1)                  (2)         
------------------------------------------------------------------------
return_1                        0.05890 (0.03802)  -1.10413*** (0.20140)
Constant                       0.17963** (0.08074) 4.65650*** (0.42768) 
------------------------------------------------------------------------
Observations                           689                  689         
R2                                   0.00348              0.04191       
Adjusted R2                          0.00203              0.04052       
Residual Std. Error (df = 687)       2.11040             11.17847       
F Statistic (df = 1; 687)            2.39946            30.05460***     
Note:                                        *p<0.

In [81]:
data("nyse")

lm.12.8 <- lm(return ~ return_1, data = nyse)
summary(lm.12.8)


Call:
lm(formula = return ~ return_1, data = nyse)

Residuals:
    Min      1Q  Median      3Q     Max 
-15.261  -1.302   0.098   1.316   8.065 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)  
(Intercept)  0.17963    0.08074   2.225   0.0264 *
return_1     0.05890    0.03802   1.549   0.1218  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.11 on 687 degrees of freedom
  (2 observations deleted due to missingness)
Multiple R-squared:  0.003481,	Adjusted R-squared:  0.00203 
F-statistic: 2.399 on 1 and 687 DF,  p-value: 0.1218


In [82]:
res <- lm.12.8$residuals^2

lm.12.8.bp <- lm(res ~ return_1, data = na.omit(nyse))
summary(lm.12.8.bp)


Call:
lm(formula = res ~ return_1, data = na.omit(nyse))

Residuals:
    Min      1Q  Median      3Q     Max 
 -9.689  -3.929  -2.021   0.960 223.730 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   4.6565     0.4277  10.888  < 2e-16 ***
return_1     -1.1041     0.2014  -5.482  5.9e-08 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 11.18 on 687 degrees of freedom
Multiple R-squared:  0.04191,	Adjusted R-squared:  0.04052 
F-statistic: 30.05 on 1 and 687 DF,  p-value: 5.905e-08


# Example 12.9: ARCH in Stock Returns
$$\hat{\mu^2_t} = \beta_0 + \hat{\mu^2_{t-1}} + residual_t$$

We still have return_mu in the working environment so we can use it to create $\hat{\mu^2_t}$, (mu2_hat) and $\hat{\mu^2_{t-1}}$ (mu2_hat_1). Notice the use R's matrix subset operations to perform the lag operation. We drop the first observation of mu2_hat and squared the results. Next, we remove the last observation of mu2_hat_1 using the subtraction operator combined with a call to the NROW function on return_mu. Now, both contain $688$ observations and we can estimate a standard linear model.

In [12]:
mu2_hat  <- return_mu[-1]^2
mu2_hat_1 <- return_mu[-NROW(return_mu)]^2
arch_model <- lm(mu2_hat ~ mu2_hat_1)
stargazer(type = "text",arch_model, single.row = TRUE, header = FALSE, digits=5)


                        Dependent variable:    
                    ---------------------------
                              mu2_hat          
-----------------------------------------------
mu2_hat_1              0.33706*** (0.03595)    
Constant               2.94743*** (0.44023)    
-----------------------------------------------
Observations                    688            
R2                            0.11361          
Adjusted R2                   0.11231          
Residual Std. Error     10.75907 (df = 686)    
F Statistic          87.92263*** (df = 1; 686) 
Note:               *p<0.1; **p<0.05; ***p<0.01


In [83]:
data("nyse")

lm.12.9 <- lm(return ~ return_1, data = nyse)
summary(lm.12.9)


Call:
lm(formula = return ~ return_1, data = nyse)

Residuals:
    Min      1Q  Median      3Q     Max 
-15.261  -1.302   0.098   1.316   8.065 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)  
(Intercept)  0.17963    0.08074   2.225   0.0264 *
return_1     0.05890    0.03802   1.549   0.1218  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.11 on 687 degrees of freedom
  (2 observations deleted due to missingness)
Multiple R-squared:  0.003481,	Adjusted R-squared:  0.00203 
F-statistic: 2.399 on 1 and 687 DF,  p-value: 0.1218


In [84]:
u <- lm.12.8$residuals
u_1 <- c(NA, u[-length(u)])
usq <- lm.12.8$residuals^2
usq_1 <- c(NA, usq[-length(usq)])

lm.12.9.arch <- lm(usq ~ usq_1)
summary(lm.12.9.arch)



Call:
lm(formula = usq ~ usq_1)

Residuals:
    Min      1Q  Median      3Q     Max 
-23.337  -3.292  -2.157   0.556 223.981 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  2.94743    0.44023   6.695 4.49e-11 ***
usq_1        0.33706    0.03595   9.377  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 10.76 on 686 degrees of freedom
  (1 observation deleted due to missingness)
Multiple R-squared:  0.1136,	Adjusted R-squared:  0.1123 
F-statistic: 87.92 on 1 and 686 DF,  p-value: < 2.2e-16


In [85]:
lm.12.9.3 <- lm(u ~ u_1)
summary(lm.12.9.3)


Call:
lm(formula = u ~ u_1)

Residuals:
     Min       1Q   Median       3Q      Max 
-15.2538  -1.3024   0.0912   1.3217   8.0613 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)
(Intercept) -0.001171   0.080508  -0.015    0.988
u_1          0.001405   0.038177   0.037    0.971

Residual standard error: 2.112 on 686 degrees of freedom
  (1 observation deleted due to missingness)
Multiple R-squared:  1.974e-06,	Adjusted R-squared:  -0.001456 
F-statistic: 0.001354 on 1 and 686 DF,  p-value: 0.9707
